lib/game.rb

```diff
@@ -20,39 +18,29 @@ def check_is_valid_human_move(user_input)
     valid_move = board.valid_move?(user_input)
     while valid_move == false
       ui.prompt_invalid_space
-      user_input = ui.get_user_input
+      user_input = ui.return_choice_integer_between(1, 9)
       valid_move = board.valid_move?(user_input)
     end
     user_input
   end
 
-  def check_is_valid_computer_move(computer_input)
-    valid_move = board.valid_move?(computer_input)
-    while valid_move == false
-      computer_input = rand(1..9)
-      valid_move = board.valid_move?(computer_input)
-    end
-    computer_input
-  end
-
   def run_game_loop
     loop do
-      turn(@player_one)
-      break if end_move?(@player_one) == true
+      turn(board: board, player: player_one)
+      break if end_move?(player_one) == true
 
-      turn(@player_two)
-      break if end_move?(@player_two) == true
+      turn(board: board, player: player_two)
+      break if end_move?(player_two) == true
     end
   end
 
-  def turn(player)
-    ui.turn_start(board)
+  def turn(board:, player:)
+    ui.turn_start(board, player)
     if player.is_computer == true
-      user_input = player.random_move
-      user_input = check_is_valid_computer_move(user_input)
+      user_input = player.move(maximizing_player: player, board: board)
```

Comment:
`maximizing_player` suggests your internals for the _unbeatable_ computer are leaking into your interface - presumably this isn't required for your non-unbeatable computer?

how might you lose the conditional and have a API where you can have something like, `player.move(board)` so you can use the open/closed principle?

Reply:
I talked with Pedro about this. I think this is a great idea that would decouple the functions and make the interface much better.